File: key-management/java/pkcs11/RSAGenerate.java
package com.unboundtech.key_management.pkcs11;

import com.dyadicsec.cryptoki.*;

public class RSAGenerate
{
  /***************
   *
   * @param args - if default user password is non-empty, provide it as first arg
   * @throws Exception
   */
  public static void main(String[] args) throws Exception
  {
    char[] pwd = null;
    if (args.length > 0) pwd = args[0].toCharArray();

    int slotId = 0;
    byte[] keyName = "TEST RSA KEY".getBytes("UTF-8");
    Library.C_Initialize();

    // Open PKCS#11 session
    System.out.println("Open Session and login with default user");
    CK_SESSION_HANDLE session = Library.C_OpenSession(slotId, CK.CKF_RW_SESSION | CK.CKF_SERIAL_SESSION);
    Library.C_Login(session, CK.CKU_USER, pwd); // optional

    // Generate key pair
    System.out.println("Generate key pair");
    int[] keyHandles = Library.C_GenerateKeyPair(session,
      new CK_MECHANISM(CK.CKM_RSA_PKCS_KEY_PAIR_GEN), // RSA generation mechanism
      new CK_ATTRIBUTE[]
      {
        new CK_ATTRIBUTE(CK.CKA_TOKEN, false),
        new CK_ATTRIBUTE(CK.CKA_MODULUS_BITS, 2048), // RSA key size
      },
      new CK_ATTRIBUTE[]
      {
        new CK_ATTRIBUTE(CK.CKA_TOKEN, true),
        new CK_ATTRIBUTE(CK.CKA_ID, keyName),
      });
    assert keyHandles.length==2;

    // Find the RSA private key
    System.out.println("Find the RSA private key");
    Library.C_FindObjectsInit(session, new CK_ATTRIBUTE[]
      {
        new CK_ATTRIBUTE(CK.CKA_TOKEN, true),
        new CK_ATTRIBUTE(CK.CKA_CLASS, CK.CKO_PRIVATE_KEY),
        new CK_ATTRIBUTE(CK.CKA_KEY_TYPE, CK.CKK_RSA),
        new CK_ATTRIBUTE(CK.CKA_ID, keyName),
      });
    int[] foundKeyHandles = Library.C_FindObjects(session, 1);
    Library.C_FindObjectsFinal(session);
    assert foundKeyHandles.length==1;
    System.out.println("Key found - success");

    // Close PKCS#11 session
    System.out.println("Close PKCS#11 session");
    Library.C_CloseSession(session);
    Library.C_Finalize();
  }
}

File: key-management/pkcs11/rsa_generate.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "dy_pkcs11.h"

void halt(CK_RV rv)
{
	printf("Error %08lx\n", rv);
	exit(-1);
}

int main(int argc, char *argv[])
{
	CK_SLOT_ID slot_id = 0; // default slot ID
	CK_SESSION_HANDLE hSession = 0;
	CK_RV rv;

	rv = C_Initialize(NULL);
	if (rv != CKR_OK)
		halt(rv);
	// Open PKCS#11 session
	rv = C_OpenSession(slot_id, CKF_SERIAL_SESSION | CKF_RW_SESSION, 0, 0, &hSession);
	if (rv != CKR_OK)
		halt(rv);

	char password[] = ""; // ------ set your password here -------
	rv = C_Login(hSession, CKU_USER, CK_CHAR_PTR(password), CK_ULONG(strlen(password)));
	if (rv != CKR_OK)
		halt(rv);

	// RSA generation mechanism
	CK_MECHANISM rsa_gen = {CKM_RSA_PKCS_KEY_PAIR_GEN, NULL, 0};

	// RSA key size
	CK_ULONG modulus_bits = 2048;

	CK_BBOOL ck_false = CK_FALSE;
	CK_BBOOL ck_true = CK_TRUE;
	char key_name[] = "TEST RSA KEY";

	CK_ATTRIBUTE pub_template[] =
		{
			{CKA_TOKEN, &ck_false, sizeof(CK_BBOOL)},
			{CKA_MODULUS_BITS, &modulus_bits, sizeof(CK_ULONG)},
		};

	CK_ATTRIBUTE prv_template[] =
		{
			{CKA_TOKEN, &ck_true, sizeof(CK_BBOOL)},
			{CKA_ID, key_name, sizeof(key_name) - 1},
		};

	// Generate key pair
	CK_OBJECT_HANDLE pub_key = 0, prv_key = 0;
	rv = C_GenerateKeyPair(hSession, &rsa_gen,
						   pub_template, sizeof(pub_template) / sizeof(CK_ATTRIBUTE),
						   prv_template, sizeof(prv_template) / sizeof(CK_ATTRIBUTE),
						   &pub_key, &prv_key);
	if (rv != CKR_OK)
		halt(rv);

	// Find RSA key type
	CK_ULONG cko_private_key = CKO_PRIVATE_KEY;
	CK_ULONG ckk_rsa = CKK_RSA;
	CK_ATTRIBUTE find_template[] =
		{
			{CKA_TOKEN, &ck_true, sizeof(CK_BBOOL)},
			{CKA_CLASS, &cko_private_key, sizeof(CK_ULONG)},
			{CKA_KEY_TYPE, &ckk_rsa, sizeof(CK_ULONG)},
			{CKA_ID, key_name, sizeof(key_name) - 1},
		};
	rv = C_FindObjectsInit(hSession, find_template, sizeof(find_template) / sizeof(CK_ATTRIBUTE));
	if (rv != CKR_OK)
		halt(rv);
	CK_ULONG found = 0;
	rv = C_FindObjects(hSession, &prv_key, 1, &found);
	if (rv != CKR_OK)
		halt(rv);
	rv = C_FindObjectsFinal(hSession);
	if (rv != CKR_OK)
		halt(rv);

	if (found != 1)
		printf("Error: Key not found");

	// Close PKCS#11 session
	C_CloseSession(hSession);
	C_Finalize(NULL);
}
